File: reader.lua
local function defreader__macro(_x, ...)
  local char = _x[1]
  local s = _x[2]
  local _r = unstash({...})
  local body = cut(_r, 0)
  return {"=", {"get", "read-table", char}, join({"fn", {s}}, body)}
end
setenv("defreader", stash33({["macro"] = defreader__macro}))
local delimiters = {["["] = true, [")"] = true, ["}"] = true, [";"] = true, ["]"] = true, ["\n"] = true, ["\r"] = true, ["{"] = true, ["("] = true}
local whitespace = {["\t"] = true, ["\r"] = true, [" "] = true, ["\n"] = true}
local function stream(str, more)
  return {["len"] = #(str), ["pos"] = 0, ["more"] = more, ["string"] = str}
end
local function peek_char(s)
  local _len = s.len
  local pos = s.pos
  local string = s.string
  if pos < _len then
    return char(string, pos)
  end
end
local function read_char(s)
  local c = peek_char(s)
  if c then
    s.pos = s.pos + 1
    return c
  end
end
local function skip_non_code(s)
  local any63 = nil
  while true do
    local c = peek_char(s)
    if c == nil then
      break
    else
      if whitespace[c] then
        read_char(s)
      else
        if c == ";" then
          while c and not( c == "\n") do
            c = read_char(s)
          end
          skip_non_code(s)
        else
          break
        end
      end
    end
    any63 = true
  end
  return any63
end
local read_table = {}
local eof = {}
local function read(s)
  skip_non_code(s)
  local c = peek_char(s)
  if c == nil then
    return eof
  else
    return (read_table[c] or read_table[""])(s)
  end
end
local function read_all(s)
  if type(s) == "string" then
    s = stream(s)
  end
  local l = {}
  while true do
    local form = read(s)
    if form == eof then
      break
    end
    add(l, form)
  end
  return l
end
local function read_string(str, more)
  local x = read(stream(str, more))
  if not( x == eof) then
    return x
  end
end
local function key63(atom)
  return type(atom) == "string" and #(atom) > 1 and char(atom, #(atom) - 1) == ":"
end
local function flag63(atom)
  return type(atom) == "string" and #(atom) > 1 and char(atom, 0) == ":"
end
local function expected(s, c)
  local pos = s.pos
  local more = s.more
  local _id6 = more
  local _e1
  if _id6 then
    _e1 = _id6
  else
    error("Expected " .. c .. " at " .. pos)
    _e1 = nil
  end
  return _e1
end
local function wrap(s, x)
  local y = read(s)
  if y == s.more then
    return y
  else
    return {x, y}
  end
end
local function maybe_number(str)
  if number_code63(code(str, #(str) - 1)) then
    return number(str)
  end
end
local function real63(x)
  return type(x) == "number" and not nan63(x) and not inf63(x)
end
local function valid_access63(str)
  return #(str) > 2 and not( "." == char(str, 0)) and not( "." == char(str, #(str) - 1)) and not search(str, "..")
end
local function parse_index(a, b)
  local n = number(a)
  if n == nil then
    return {"get", b, {"quote", a}}
  else
    return {"at", b, n}
  end
end
local function parse_access(str, prev)
  local _e2
  if prev then
    _e2 = {prev}
  else
    _e2 = {}
  end
  local parts = _e2
  return reduce(parse_index, rev(join(parts, split(str, "."))))
end
local function read_atom(s, basic63)
  local str = ""
  local dot63 = false
  while true do
    local c = peek_char(s)
    if not c or whitespace[c] or delimiters[c] then
      break
    end
    if c == "." then
      dot63 = true
    end
    if c == "\\" then
      read_char(s)
    end
    str = str .. read_char(s)
  end
  if str == "true" then
    return true
  else
    if str == "false" then
      return false
    else
      if str == "nan" then
        return nan
      else
        if str == "-nan" then
          return nan
        else
          if str == "inf" then
            return inf
          else
            if str == "-inf" then
              return -inf
            else
              if basic63 then
                return str
              else
                local n = maybe_number(str)
                if real63(n) then
                  return n
                else
                  if dot63 and valid_access63(str) then
                    return parse_access(str)
                  else
                    return str
                  end
                end
              end
            end
          end
        end
      end
    end
  end
end
local function read_list(s, ending)
  read_char(s)
  local r = nil
  local l = {}
  while r == nil do
    skip_non_code(s)
    local c = peek_char(s)
    if c == ending then
      read_char(s)
      r = l
    else
      if c == nil then
        r = expected(s, ending)
      else
        local x = read(s)
        if key63(x) then
          local k = clip(x, 0, #(x) - 1)
          local v = read(s)
          l[k] = v
        else
          if flag63(x) then
            l[clip(x, 1)] = true
          else
            add(l, x)
          end
        end
      end
    end
  end
  return r
end
local function read_next(s, prev, ws63)
  local _e = peek_char(s)
  if "." == _e then
    read_char(s)
    skip_non_code(s)
    if not peek_char(s) then
      return s.more or eof
    else
      local x = read_atom(s, true)
      if x == eof or x == s.more then
        return x
      else
        return read_next(s, parse_access(x, prev))
      end
    end
  else
    if "(" == _e then
      if ws63 then
        return prev
      else
        local _x11 = read_list(s, ")")
        if _x11 == s.more then
          return _x11
        else
          return read_next(s, join({prev}, _x11), skip_non_code(s))
        end
      end
    else
      return prev
    end
  end
end
read_table[""] = function (s)
  return read_next(s, read_atom(s))
end
read_table["("] = function (s)
  return read_next(s, read_list(s, ")"), skip_non_code(s))
end
read_table[")"] = function (s)
  error("Unexpected ) at " .. s.pos)
end
function ontree(f, l, ...)
  local _r23 = unstash({...})
  local skip = _r23.skip
  if not( skip and skip(l)) then
    local y = f(l)
    if y then
      return y
    end
    if not not( type(l) == "table") then
      local _l = l
      local _i = nil
      for _i in next, _l do
        local x = _l[_i]
        local _y = ontree(f, x, stash33({["skip"] = skip}))
        if _y then
          return _y
        end
      end
    end
  end
end
function hd_is63(l, val)
  return type(l) == "table" and l[1] == val
end
local function _37fn__macro(body)
  local n = -1
  local l = {}
  local any63 = nil
  ontree(function (_)
    if type(_) == "string" and #(_) <= 2 and code(_, 0) == 95 then
      any63 = true
      local c = code(_, 1)
      if c and c >= 48 and c <= 57 then
        while n < c - 48 do
          n = n + 1
          add(l, "_" .. chr(48 + n))
        end
      end
      return nil
    end
  end, body, stash33({["skip"] = function (_)
    return hd_is63(_, "%fn")
  end}))
  if any63 and #(l) == 0 then
    add(l, "_")
  end
  return {"fn", l, body}
end
setenv("%fn", stash33({["macro"] = _37fn__macro}))
read_table["["] = function (s)
  local x = read_list(s, "]")
  if x == s.more then
    return x
  else
    return read_next(s, {"%fn", x}, skip_non_code(s))
  end
end
read_table["]"] = function (s)
  error("Unexpected ] at " .. s.pos)
end
read_table["{"] = function (s)
  local x = read_list(s, "}")
  if x == s.more then
    return x
  else
    local e = x[1]
    x = cut(x, 1)
    while #(x) > 1 do
      local op = x[1]
      local a = x[2]
      local bs = cut(x, 2)
      e = {op, e, a}
      x = bs
    end
    return read_next(s, e, skip_non_code(s))
  end
end
read_table["}"] = function (s)
  error("Unexpected } at " .. s.pos)
end
read_table["\""] = function (s)
  read_char(s)
  local r = nil
  local str = "\""
  while r == nil do
    local c = peek_char(s)
    if c == "\"" then
      r = str .. read_char(s)
    else
      if c == nil then
        r = expected(s, "\"")
      else
        if c == "\\" then
          str = str .. read_char(s)
        end
        str = str .. read_char(s)
      end
    end
  end
  return r
end
read_table["|"] = function (s)
  read_char(s)
  local r = nil
  local str = "|"
  while r == nil do
    local c = peek_char(s)
    if c == "|" then
      r = str .. read_char(s)
    else
      if c == nil then
        r = expected(s, "|")
      else
        str = str .. read_char(s)
      end
    end
  end
  return r
end
read_table["'"] = function (s)
  read_char(s)
  return wrap(s, "quote")
end
read_table["`"] = function (s)
  read_char(s)
  return wrap(s, "quasiquote")
end
read_table[","] = function (s)
  read_char(s)
  if peek_char(s) == "@" then
    read_char(s)
    return wrap(s, "unquote-splicing")
  else
    return wrap(s, "unquote")
  end
end
return {["read-string"] = read_string, ["read-all"] = read_all, ["read"] = read, ["read-table"] = read_table, ["stream"] = stream}

File: reader.js
var defreader__macro = function (_x) {
  var _char = _x[0];
  var _s = _x[1];
  var __r = unstash(Array.prototype.slice.call(arguments, 1));
  var _body = cut(__r, 0);
  return ["=", ["get", "read-table", _char], join(["fn", [_s]], _body)];
};
setenv("defreader", stash33({["macro"]: defreader__macro}));
var delimiters = {["["]: true, [")"]: true, ["}"]: true, [";"]: true, ["]"]: true, ["\n"]: true, ["\r"]: true, ["{"]: true, ["("]: true};
var whitespace = {["\t"]: true, ["\r"]: true, [" "]: true, ["\n"]: true};
var stream = function (str, more) {
  return {["len"]: str.length || 0, ["pos"]: 0, ["more"]: more, ["string"]: str};
};
var peek_char = function (s) {
  var _len = s.len;
  var _pos = s.pos;
  var _string = s.string;
  if (_pos < _len) {
    return char(_string, _pos);
  }
};
var read_char = function (s) {
  var _c = peek_char(s);
  if (_c) {
    s.pos = s.pos + 1;
    return _c;
  }
};
var skip_non_code = function (s) {
  var _any63 = undefined;
  while (true) {
    var _c1 = peek_char(s);
    if (typeof(_c1) === "undefined" || _c1 === null) {
      break;
    } else {
      if (whitespace[_c1]) {
        read_char(s);
      } else {
        if (_c1 === ";") {
          while (_c1 && !( _c1 === "\n")) {
            _c1 = read_char(s);
          }
          skip_non_code(s);
        } else {
          break;
        }
      }
    }
    _any63 = true;
  }
  return _any63;
};
var read_table = {};
var eof = {};
var read = function (s) {
  skip_non_code(s);
  var _c2 = peek_char(s);
  if (typeof(_c2) === "undefined" || _c2 === null) {
    return eof;
  } else {
    return (read_table[_c2] || read_table[""])(s);
  }
};
var read_all = function (s) {
  if (typeof(s) === "string") {
    s = stream(s);
  }
  var _l = [];
  while (true) {
    var _form = read(s);
    if (_form === eof) {
      break;
    }
    add(_l, _form);
  }
  return _l;
};
var read_string = function (str, more) {
  var _x5 = read(stream(str, more));
  if (!( _x5 === eof)) {
    return _x5;
  }
};
var key63 = function (atom) {
  return typeof(atom) === "string" && (atom.length || 0) > 1 && char(atom, (atom.length || 0) - 1) === ":";
};
var flag63 = function (atom) {
  return typeof(atom) === "string" && (atom.length || 0) > 1 && char(atom, 0) === ":";
};
var expected = function (s, c) {
  var _pos1 = s.pos;
  var _more = s.more;
  var _id6 = _more;
  var _e2;
  if (_id6) {
    _e2 = _id6;
  } else {
    throw new Error("Expected " + c + " at " + _pos1);
    _e2 = undefined;
  }
  return _e2;
};
var wrap = function (s, x) {
  var _y = read(s);
  if (_y === s.more) {
    return _y;
  } else {
    return [x, _y];
  }
};
var maybe_number = function (str) {
  if (number_code63(code(str, (str.length || 0) - 1))) {
    return number(str);
  }
};
var real63 = function (x) {
  return typeof(x) === "number" && ! nan63(x) && ! inf63(x);
};
var valid_access63 = function (str) {
  return (str.length || 0) > 2 && !( "." === char(str, 0)) && !( "." === char(str, (str.length || 0) - 1)) && ! search(str, "..");
};
var parse_index = function (a, b) {
  var _n = number(a);
  if (typeof(_n) === "undefined" || _n === null) {
    return ["get", b, ["quote", a]];
  } else {
    return ["at", b, _n];
  }
};
var parse_access = function (str, prev) {
  var _e3;
  if (prev) {
    _e3 = [prev];
  } else {
    _e3 = [];
  }
  var _parts = _e3;
  return reduce(parse_index, rev(join(_parts, split(str, "."))));
};
var read_atom = function (s, basic63) {
  var _str = "";
  var _dot63 = false;
  while (true) {
    var _c3 = peek_char(s);
    if (! _c3 || whitespace[_c3] || delimiters[_c3]) {
      break;
    }
    if (_c3 === ".") {
      _dot63 = true;
    }
    if (_c3 === "\\") {
      read_char(s);
    }
    _str = _str + read_char(s);
  }
  if (_str === "true") {
    return true;
  } else {
    if (_str === "false") {
      return false;
    } else {
      if (_str === "nan") {
        return nan;
      } else {
        if (_str === "-nan") {
          return nan;
        } else {
          if (_str === "inf") {
            return inf;
          } else {
            if (_str === "-inf") {
              return -inf;
            } else {
              if (basic63) {
                return _str;
              } else {
                var _n1 = maybe_number(_str);
                if (real63(_n1)) {
                  return _n1;
                } else {
                  if (_dot63 && valid_access63(_str)) {
                    return parse_access(_str);
                  } else {
                    return _str;
                  }
                }
              }
            }
          }
        }
      }
    }
  }
};
var read_list = function (s, ending) {
  read_char(s);
  var _r19 = undefined;
  var _l1 = [];
  while (typeof(_r19) === "undefined" || _r19 === null) {
    skip_non_code(s);
    var _c4 = peek_char(s);
    if (_c4 === ending) {
      read_char(s);
      _r19 = _l1;
    } else {
      if (typeof(_c4) === "undefined" || _c4 === null) {
        _r19 = expected(s, ending);
      } else {
        var _x11 = read(s);
        if (key63(_x11)) {
          var _k = clip(_x11, 0, (_x11.length || 0) - 1);
          var _v = read(s);
          _l1[_k] = _v;
        } else {
          if (flag63(_x11)) {
            _l1[clip(_x11, 1)] = true;
          } else {
            add(_l1, _x11);
          }
        }
      }
    }
  }
  return _r19;
};
var read_next = function (s, prev, ws63) {
  var __e = peek_char(s);
  if ("." === __e) {
    read_char(s);
    skip_non_code(s);
    if (! peek_char(s)) {
      return s.more || eof;
    } else {
      var _x12 = read_atom(s, true);
      if (_x12 === eof || _x12 === s.more) {
        return _x12;
      } else {
        return read_next(s, parse_access(_x12, prev));
      }
    }
  } else {
    if ("(" === __e) {
      if (ws63) {
        return prev;
      } else {
        var _x13 = read_list(s, ")");
        if (_x13 === s.more) {
          return _x13;
        } else {
          return read_next(s, join([prev], _x13), skip_non_code(s));
        }
      }
    } else {
      return prev;
    }
  }
};
read_table[""] = function (s) {
  return read_next(s, read_atom(s));
};
read_table["("] = function (s) {
  return read_next(s, read_list(s, ")"), skip_non_code(s));
};
read_table[")"] = function (s) {
  throw new Error("Unexpected ) at " + s.pos);
};
ontree = function (f, l) {
  var __r24 = unstash(Array.prototype.slice.call(arguments, 2));
  var _skip = __r24.skip;
  if (!( _skip && _skip(l))) {
    var _y1 = f(l);
    if (_y1) {
      return _y1;
    }
    if (! !( typeof(l) === "object")) {
      var __l2 = l;
      var __i = undefined;
      for (__i in __l2) {
        var _x15 = __l2[__i];
        var _e4;
        if (numeric63(__i)) {
          _e4 = parseInt(__i);
        } else {
          _e4 = __i;
        }
        var __i1 = _e4;
        var _y2 = ontree(f, _x15, stash33({["skip"]: _skip}));
        if (_y2) {
          return _y2;
        }
      }
    }
  }
};
hd_is63 = function (l, val) {
  return typeof(l) === "object" && l[0] === val;
};
var _37fn__macro = function (body) {
  var _n3 = -1;
  var _l3 = [];
  var _any631 = undefined;
  ontree(function (_) {
    if (typeof(_) === "string" && (_.length || 0) <= 2 && code(_, 0) === 95) {
      _any631 = true;
      var c = code(_, 1);
      if (c && c >= 48 && c <= 57) {
        while (_n3 < c - 48) {
          _n3 = _n3 + 1;
          add(_l3, "_" + chr(48 + _n3));
        }
      }
      return undefined;
    }
  }, body, stash33({["skip"]: function (_) {
    return hd_is63(_, "%fn");
  }}));
  if (_any631 && (_l3.length || 0) === 0) {
    add(_l3, "_");
  }
  return ["fn", _l3, body];
};
setenv("%fn", stash33({["macro"]: _37fn__macro}));
read_table["["] = function (s) {
  var _x17 = read_list(s, "]");
  if (_x17 === s.more) {
    return _x17;
  } else {
    return read_next(s, ["%fn", _x17], skip_non_code(s));
  }
};
read_table["]"] = function (s) {
  throw new Error("Unexpected ] at " + s.pos);
};
read_table["{"] = function (s) {
  var _x19 = read_list(s, "}");
  if (_x19 === s.more) {
    return _x19;
  } else {
    var _e1 = _x19[0];
    _x19 = cut(_x19, 1);
    while ((_x19.length || 0) > 1) {
      var _op = _x19[0];
      var _a = _x19[1];
      var _bs = cut(_x19, 2);
      _e1 = [_op, _e1, _a];
      _x19 = _bs;
    }
    return read_next(s, _e1, skip_non_code(s));
  }
};
read_table["}"] = function (s) {
  throw new Error("Unexpected } at " + s.pos);
};
read_table["\""] = function (s) {
  read_char(s);
  var _r34 = undefined;
  var _str1 = "\"";
  while (typeof(_r34) === "undefined" || _r34 === null) {
    var _c5 = peek_char(s);
    if (_c5 === "\"") {
      _r34 = _str1 + read_char(s);
    } else {
      if (typeof(_c5) === "undefined" || _c5 === null) {
        _r34 = expected(s, "\"");
      } else {
        if (_c5 === "\\") {
          _str1 = _str1 + read_char(s);
        }
        _str1 = _str1 + read_char(s);
      }
    }
  }
  return _r34;
};
read_table["|"] = function (s) {
  read_char(s);
  var _r36 = undefined;
  var _str2 = "|";
  while (typeof(_r36) === "undefined" || _r36 === null) {
    var _c6 = peek_char(s);
    if (_c6 === "|") {
      _r36 = _str2 + read_char(s);
    } else {
      if (typeof(_c6) === "undefined" || _c6 === null) {
        _r36 = expected(s, "|");
      } else {
        _str2 = _str2 + read_char(s);
      }
    }
  }
  return _r36;
};
read_table["'"] = function (s) {
  read_char(s);
  return wrap(s, "quote");
};
read_table["`"] = function (s) {
  read_char(s);
  return wrap(s, "quasiquote");
};
read_table[","] = function (s) {
  read_char(s);
  if (peek_char(s) === "@") {
    read_char(s);
    return wrap(s, "unquote-splicing");
  } else {
    return wrap(s, "unquote");
  }
};
exports.stream = stream;
exports.read = read;
exports["read-all"] = read_all;
exports["read-string"] = read_string;
exports["read-table"] = read_table;
